### ISSUE TITLE
Sifnode Load Testing for Peggy 2.0

### ISSUE THREAD

**Issue by pandaring2you** [Sifchain/sifnode#2520]
Hey @Jedi2002 - The peggy team was going through the load testing plan for Peggy 2.0 and realized that Sifnode will likely want to conduct load testing given that Peggy 2.0 will be removing the whitelist and potentially thousands of new tokens will be in our system. Have you considered what type of load testing we'll need to conduct?

One potential scenario:
- Create a wallet with 1K, 10K, 100K, different denoms

@smartyalgo @banshee - feel free to comment if you think of additional scenarios sifnode should be testing

cc: @marshsif

**Comment by pandaring2you** [comment]
Testing needed prior to peggy 2.0. Depending on results, need to determine if it blocks peggy 2.0 launch

**Comment by pandaring2you** [comment]
hey @Jedi2002 - can you give this a look and let us know timing?

**Comment by pandaring2you** [comment]
@khdegraaf We will need a partner on the testing team to suggest scenarios to test for to work with @iverc @joneskm . Do you have a suggestion on who would be best to work on this?

Known issue: Adding thousands of tokens in the registry, increasing the cost of queries/operations. 

Potential solutions: 
- Optimize queries/operations
- Add cost to import tokens. 

@pandaring2you - Write up doc on potential token spam attacks, get feedback from testing team and peggy. Then organize meeting to draw up solutions.

**Comment by pandaring2you** [comment]
@jzvikart - to assess scenarios - order them by severity. determine severity and discuss solutions with sifnode/peggy teams

**Comment by jzvikart** [comment]
We can get a large number of tokens in a sif wallet in two ways:
- by importing them from ERC20 tokens
- by creating them in sifnoded genesis (requires parametrization of environment + starting environment from within the test itself)

Because those tokens are different (one are cosmos native and others are not), we need to test them with two different scenarios. For now, we will focus on EVM-native tokens.

Next steps:
- Expand load testing framework to deploy and mint an arbitrary number of ERC20 tokens and send them to a sif wallet
- Run it with 10, 100, 1k, 10k, 100k and see how many tokens can be realistically created in ~30 mins
- Perform some basic tests on this wallet such as checking balance, sending some tokens back to Ethereum, sending some tokens to another sif wallet
- Discuss with @iverc how to measure/detect potential performance issues with Sifnode involving a large number of 
denoms in a wallet (performance must be O(1) or O(n) at worst) - might require profiling instrumentation on sifnode side
- If necessary, research how we could create snapshots of EVM + sifnode environments (hardhat forking, Infura, geth etc.)

**Comment by jzvikart** [comment]
I was able to deploy 5000 ERC20 tokens and send them to sif wallet and see them in the balance. This took about 3 hours on a 4CPU VM with 24G RAM. So far there were no surprises, but I will add some more logic to the test if the balances are actually working as expected (transfer, burn etc.)

With current approach, anything over 5000 denoms is problematic. Hardhat's performance deteriorates drastically after certain time/number of blocks (by factor of 200). I tried running the test with 10000 denoms, but the test did not complete even after 12 hours and the rate of ethereum calls dropped to <0.1/s. Hardhat CPU was constantly about 150%, with around 6GB non-shared memory usage. If we want to go over that, we should probably switch to geth.

Regarding sifnoded performance - I tried to contact @iverc but could not reach him.

**Comment by pandaring2you** [comment]
AI: Use geth for now, will try to push 10K denoms. First issue to resolve is timing out when increasing denoms

**Comment by jzvikart** [comment]
Results for 5000 denoms on 4 CPU with 24G RAM and using hardhat:

```
Number of tokens: 5000
batch_deploy_erc20_tokens(): 274.27 s, 18.23 items/s
batch_mint_erc20_tokens(): 399.68 s, 12.51 items/s
batch_approve_and_lock_erc20_tokens(): 1572.43 s, 3.18 items/s
wait_for_sif_balance_change(): 0.61 s, 8173.75 items/s
Cost of approve+lock: 210544.61 gwei
Total for setup (mint+approve+lock): 2248.07 s, 2.22 items/s
Average tx_send time for fat_sif_wallet: 4.93 s
Average tx_burn time for fat_sif_wallet: 56.52 s
Average tx_send time for slim_sif_wallet: 5.10 s
Average tx_burn time for slim_sif_wallet: 59.52 s
Relative fat/slim speed for tx_send: 0.97
Relative fat/slim speed for tx_burn: 0.95
Total test time: 3597.30 s
```

**Comment by iverc** [comment]
@jzvikart thanks for sharing test results. it would be good to have same report for 1,000 tokens on the same CPU/RAM to compare

**Comment by pandaring2you** [comment]
@jzvikart - is this the feedback you needed from @iverc . if not, @iverc, please get on the phone with jure and unblock him.

**Comment by jzvikart** [comment]
@pandaring2you No, I would need @iverc to discuss how to test liquidity pools and internal sifnoded performance.

**Comment by jzvikart** [comment]
Results for 1000 denoms on 4 CPU with 24G RAM and using hardhat:

```
Number of tokens: 1000
batch_deploy_erc20_tokens(): 32.09 s, 31.16 items/s
batch_mint_erc20_tokens(): 36.67 s, 27.27 items/s
batch_approve_and_lock_erc20_tokens(): 104.45 s, 9.57 items/s
wait_for_sif_balance_change(): 11.28 s, 88.67 items/s
Cost of approve+lock: 210558.39 gwei
Total for setup (mint+approve+lock): 184.75 s, 5.41 items/s
Average tx_send time for fat_sif_wallet: 5.05 s
Average tx_burn time for fat_sif_wallet: 57.13 s
Average tx_send time for slim_sif_wallet: 5.09 s
Average tx_burn time for slim_sif_wallet: 60.18 s
Relative fat/slim speed for tx_send: 0.99
Relative fat/slim speed for tx_burn: 0.95
Total test time: 1492.54 s
```

**Comment by pandaring2you** [comment]
If Ivan is unavailable, other folks on the Sifnode team might be freed up after release today. Perhaps Tim or Tanmay? Austin is on pto till tuesday.

**Comment by jzvikart** [comment]
Look here for hints about testing liquidity pools:
* `test/integration/src/py/test_liquidity_pools.py`
* `x/clp/handler_test.go`
* `x/clp/scenarios_test.go`

**Comment by pandaring2you** [comment]
Jure chatted with Ivan. Want to do performance testing for Sifnode. Need to determine plan w/ @iverc for testing scenarios and solutions.

**Comment by Jedi2002** [comment]
Goin throuigh the conversation , I am not sure why do we need to use ERC20 tokens ? 
- Cant we just mint tokens with random names , for an address . 
- Import that address to a wallet / Or use CLI 
- And then try to write tests around it ?

We might need to paginate the bank balances query

**Comment by marshsif** [comment]
just noting that w should load test both large token set and LP set to see what the upper limits of the protocol currently are, identify areas where the code has scaling limitations, and make improvements. We'll want to identify our scaling limits and potentially set a restriction once those limits are reached.

**Comment by jzvikart** [comment]
After discussing with @iverc:
- There are two ways to get a large number of tokens into a wallet: (1) through Peggy (slow, requires switching to geth instead of hardhat), and (2) through genesis (limited by size of command line / size of genesis JSON file)
- From sifnoded side of things all denoms are equivalent and it doesn't matter where they come from. We do need at least some ERC20 tokens since we also want to check the performance of the peggy2 separately from sifnoded.
- Most critical sifnoded operations are:
  - get balance
  - send tokens to another wallet
- To estimate O(x), we should run the operations for 10, 100, 1000 tokens and compare the time they take to run. For example, if `q bank balances` for 1000 tokens takes 10 times as long as for 100 tokens, and 100 times as long as for 10 tokens, we can assume that the operation is O(n).
- O(n) is OK, anything significantly more is suspicious and should be investigated.
- We should run the tests on a minimal spec machine (i.e. VM where we can set available memory/CPU)
- Current assumption is that if everything works with 100k tokens we consider the result as "pass"
- Liquidity pools are not immediate concern for "many denoms" test - no need to test them for now (reason: since we create the pools ourselves we're not going to hit the bottleneck overnight; but we might want to add a separate test for large number of pools).

@marshsif @Jedi2002 Does that sound reasonable with you? Let me know if you have any concerns and/or ideas.

**Comment by Jedi2002** [comment]
Hey, @jzvikart we need to write the data to the genesis.json to add via init-genesis.
Please see this example code: https://github.com/Sifchain/sifnode/blob/69f91616b5469bae7a2aa2f32770bfc3e400ef53/x/dispensation/genesis.go#L18
I don't think there is any limitation to adding the data through genesis unless we are talking about terabytes of data.

Send should not be a problem logically, since we specify the coin, for balances I think as long as we paginate it should run fine.

**Comment by jzvikart** [comment]
Adding denoms via `sifnoded add-genesis-account` is problematic because both shell and OS have limited maximum command line length. Practical usability limit is about 10k denoms (assuming short denom names).

Adding denoms via `genesis.json` works somewhat better. In this way I could create a wallet with 100k denoms. However, sifnoded takes significantly more time to start, being CPU-bound:
- for 10k denoms: 12s
- for 50k denoms: 222s (3.7x of 10k x 5)
- for 100k denoms: 948s (7.9x of 10k x 10)

This looks like O(n^2).

Time for `sifnoded q bank balances --limit 1000000`:
- for 10k denoms: 0.5s
- for 50k denoms: 2.3s (0.9x of 10k x 5)
- for 100k denoms: 4.7s (0.9x of 10k x 10)

This looks like O(n).

**Comment by Jedi2002** [comment]
Yeah that is correct it should. Adding via genesis is just a simple way of simulation.

I had expected the bank balances query to be slow. Could you try wrapping the query a paginated version. That would solve this problem.

**Comment by Jedi2002** [comment]
Also would be curious to know if you managed to test send TX, to see of its impacted in any way?

**Comment by Jedi2002** [comment]
We don't need to fetch all the tokens when querying the balance. We can just paginate the query.

**Comment by pandaring2you** [comment]
@jzvikart - to focus on testing
@Jedi2002 - to prioritize pagination

**Comment by pandaring2you** [comment]
@Jedi2002 - @timlind wanted to get an answer from you on whether or not we should include pagination in sifnode or extend the SDK. 
@alanrsoares - Could also use answers from FE to determine we need to build a new endpoint for the bank balances. 
@jzvikart  - Do you have an outline of your test plan for @timlind ?

**Comment by pandaring2you** [comment]
Current estimate: 0.5 weeks (once we get the above questions answered)

**Comment by jzvikart** [comment]
@pandaring2you @Jedi2002 How is pagination related to this? As far as the results of my test go, there is no problem that needs to be fixed here. The problem (if anything) is with large genesis file.

**Comment by kingpinXD** [comment]
Pagination will just add an option to query limited data, instead of querying all denoms.
It wont improve query performance which will still remain O(N) .

**Comment by Jedi2002** [comment]
Pagination will just add an option to query limited data, instead of querying all denoms.
It wont improve query performance which will still remain O(N) .

**Comment by Jedi2002** [comment]
Also @timlind @pandaring2you 
I feel for adding the migrator for bank as mentioned here https://app.zenhub.com/workspaces/current-sprint---engineering-615a2e9fe2abd5001befc7f9/issues/sifchain/sifnode/2420

we would need to create a wrapper for the sdk bank anyways , we can include paginating the query as part of the same task

**Comment by pandaring2you** [comment]
hey guys - following up on this @timlind - did you get the answers you need? @jzvikart - were you able to talk to @Jedi2002 about the pagination?

**Comment by timlind** [comment]
Yes, @Jedi2002 and I noticed the pagination is already part of the SDK so that can be considered done. 

@jzvikart are you planning on testing with a large number of pools next?

**Comment by jzvikart** [comment]
> @jzvikart are you planning on testing with a large number of pools next?

@timlind I'm not aware of any planned Peggy-related test that covers pools at this time. If you think this is something that we should tested, let's get in touch and design a test scenario and test cases. We should probably open a separate ticket for that.

**Comment by jzvikart** [comment]
About pagination - to prevent against DoS-style attacks there should probably be a built-in limit to maximum number of results per request. Also, from the point of caller it would have to be clarified what happens if the balances change between requests for individual pages.

**Comment by pandaring2you** [comment]
@pandaring2you to follow up on meeting

**Comment by timlind** [comment]
@jzvikart There is a PR up for the pagination work which can be tested https://github.com/Sifchain/sifnode/pull/2794

**Comment by jzvikart** [comment]
@timlind Thanks, I saw it, but have you also seen my questions there?

**Comment by pandaring2you** [comment]
having mining issues due to hardhat <> geth transition. tried investigating geth's development mode but does not work with ebrelayer. not sure where to go from here (we did 10K tokens with hardhat, but can't do 100K with geth) - need to discuss/decide with Peggy team. 

want to try to switch our tests to geth, but brando does not think this is a priority.

**Comment by pandaring2you** [comment]
Jure/Brandon spoke - Jure will continue to investigate, will want to conduct this test with geth as geth is production

**Comment by jzvikart** [comment]
At the moment, adjusting the block gas limit seems feasible for 100k denoms, but I'm running into limitations of geth crashing because of out-of-memory and disk filling up because of ebrelayer logs. Still investigating, I might need to shift to running the tests on AWS if my machine can't handle it.

**Comment by pandaring2you** [comment]
Want to run load tests with a large amount of real production data to test performance of sif wallets (vs running tests against small amounts of synthetic data). Test times to fetch a wallet are getting longer with increasing data.

@Jedi2002 says that we can create this data via genesis file.

**Comment by jzvikart** [comment]
Testing accounts with large number of denoms is blocked by #2858.

**Comment by jzvikart** [comment]
Results for 1000 denoms using geth:
```
Number of tokens: 1000
batch_deploy_erc20_tokens(): 18.20 s, 54.96 items/s
batch_mint_erc20_tokens(): 28.10 s, 35.59 items/s
batch_approve_and_lock_erc20_tokens(): 36.16 s, 27.65 items/s
wait_for_sif_balance_change(): 58.02 s, 17.23 items/s
Cost of approve+lock: 216210.03 gwei
Total for setup (mint+approve+lock): 140.65 s, 7.11 items/s
Average tx_send time for fat_sif_wallet: 5.13 s
Average tx_burn time for fat_sif_wallet: 55.47 s
Average tx_send time for slim_sif_wallet: 5.14 s
Average tx_burn time for slim_sif_wallet: 59.90 s
Relative fat/slim speed for tx_send: 1.00
Relative fat/slim speed for tx_burn: 0.93
Total test time: 1489.45 s
```

Results for 5000 denoms using geth:
```
Number of tokens: 5000
batch_deploy_erc20_tokens(): 92.37 s, 54.13 items/s
batch_mint_erc20_tokens(): 159.63 s, 31.32 items/s
batch_approve_and_lock_erc20_tokens(): 265.83 s, 18.81 items/s
wait_for_sif_balance_change(): 41.29 s, 121.09 items/s
Cost of approve+lock: 216193.61 gwei
Total for setup (mint+approve+lock): 559.65 s, 8.93 items/s
Average tx_send time for fat_sif_wallet: 5.00 s
Average tx_burn time for fat_sif_wallet: 60.16 s
Average tx_send time for slim_sif_wallet: 5.17 s
Average tx_burn time for slim_sif_wallet: 59.61 s
Relative fat/slim speed for tx_send: 0.97
Relative fat/slim speed for tx_burn: 1.01
Total test time: 1953.98 s
```

Results for 10000 denoms using geth:
```
Number of tokens: 10000
batch_deploy_erc20_tokens(): 197.83 s, 50.55 items/s
batch_mint_erc20_tokens(): 323.49 s, 30.91 items/s
batch_approve_and_lock_erc20_tokens(): 474.52 s, 21.07 items/s
wait_for_sif_balance_change(): 19.90 s, 502.42 items/s
Cost of approve+lock: 216195.33 gwei
Total for setup (mint+approve+lock): 1017.07 s, 9.83 items/s
Average tx_send time for fat_sif_wallet: 5.25 s
Average tx_burn time for fat_sif_wallet: 58.04 s
Average tx_send time for slim_sif_wallet: 5.27 s
Average tx_burn time for slim_sif_wallet: 59.48 s
Relative fat/slim speed for tx_send: 1.00
Relative fat/slim speed for tx_burn: 0.98
Total test time: 2393.13 s
```

**Comment by pandaring2you** [comment]
From Jure: 
The good news is that Peggy seems to be working OK. In other words, a wallet with 10k denoms is not any slower in terms of locks/burns than a wallet with 1 denom.

The bad news is that there seems to be no way to list all the balances for now. sifnoded query bank balances times out for >~50000 items. I'm assuming that this is a Cosmos issue. Created issue #2858

**Comment by jzvikart** [comment]
Next problem: after successfully locking 100k tokens on Ethereum side (with geth), only ~92600 make it to the sif wallet.

In witness.log:
```
{"L":"ERROR","T":"2022-06-04T14:53:35.130+0200","C":"relayer/ethereum.go:317","M":"failed to handle ethereum event.","errorMessage":"RPC error -32600 - Invalid Request: error reading request body: http: request body too large","S":"github.com/Sifchain/sifnode/cmd/ebrelayer/relayer.EthereumSub.CheckNonceAndProcess\n\t/.../cmd/ebrelayer/relayer/ethereum.go:317\ngithub.com/Sifchain/sifnode/cmd/ebrelayer/relayer.EthereumSub.Start\n\t/.../cmd/ebrelayer/relayer/ethereum.go:151"}
```

**Comment by jzvikart** [comment]
Created a separate issue for above: #2881

**Comment by pandaring2you** [comment]
From Jure: 
"Good news everyone!
The script finished with 100k tokens created. What's more, I was even able to read all the balances.
It took a long time though (>30 hours) - the average transaction rate is less than 1 tps. Before changes it was considerably more, somewhere in the range of 10-20 tps.
This leads me to believe that we now have to fine-tune the parameters both for reading from Ethereum and for sending to Cosmos and use the biggest possible batch sizes that will not exceed the maximum request sizes.
Also to consider: there might be a analogous case for sending transactions from Cosmos to Ethereum which we haven't addressed yet. 
Excessive logging and high CPU usage are still an issue too."

**Comment by pandaring2you** [comment]
From Jure

"In my understanding we need a better way of implementing forwarding (it's a typical producer-consumer pattern) which would maximize both receive and send buffers as much as possible without sacrificing the stability.
I communicated this to Michael and Brandon, so they are both aware of it, however I haven't checked yet if they implemented any improvements over initial fix. I expect that after we settle on a proper solution I will run another batch of load tests to check it."

**Comment by pandaring2you** [comment]
Re-ran tests - ran into a timeout issue. Attempting to resolve now. Note: CPU usage is 3-5% -seems normal. Suspect this improvement is a result of Junius's changes. Junius has a separate task to profile cpu usage - less of a priority since CPU usage is back to normal.

**Comment by pandaring2you** [comment]
@jzvikart to send link to collab doc for testcases to @pandaring2you to send doc to khdegraaf joneskm jzvikart sheokapr gzukel timlind to think through additional test cases for sifnode load testing for peggy 2.

**Comment by jzvikart** [comment]
Proposal for creating new load tests:
- Create a `.py` file in `test/load` with the name of scenario
- Write scenario as comments (`#` - style)
- Implement scenario as either main or pytest module
- Optionally, import `load_testing` module to reuse existing functionality from other load tests
- Keep the comments as the specification and documentation

**Comment by pandaring2you** [comment]
Note from @khdegraaf : Notional mentioned that block time is typically impacted by I/O and querying. CPU time is typically not a concern.
